src/xrpld/app/tx/detail/OnChainCredentials.cpp

```diff
@@ -0,0 +1,454 @@
+}
+
+// ------- DELETE --------------------------
+NotTEC
+CredentialDelete::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureOnChainCredentials))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto& tx = ctx.tx;
+    auto& j = ctx.j;
+
+    auto const optSubj = tx[~sfSubject];
+    auto const optIss = tx[~sfIssuer];
+
+    if (!optSubj && !optIss)
+    {
+        // Neither field is present, the transaction is malformed.
+        JLOG(j.trace()) << "Malformed transaction: "
+                           "Invalid Subject and Issuer fields combination.";
+        return temMALFORMED;
+    }
+
+    // Make sure that the passed account is valid.
+    if ((optSubj && !AccountID(*optSubj)) || (optIss && !AccountID(*optIss)))
+    {
+        JLOG(j.trace()) << "Malformed transaction: Subject or Issuer "
+                           "field zeroed.";
+        return temINVALID_ACCOUNT_ID;
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+CredentialDelete::preclaim(PreclaimContext const& ctx)
+{
+    AccountID const acc{ctx.tx[sfAccount]};
+
+    auto const& txCredType(ctx.tx[sfCredentialType]);
+    Blob const credType{txCredType.begin(), txCredType.end()};
+
+    auto const subj = ctx.tx[~sfSubject];
+    auto const iss = ctx.tx[~sfIssuer];
+    auto const kCred =
+        keylet::credential(subj ? *subj : acc, iss ? *iss : acc, credType);
+    auto sleCred = ctx.view.read(kCred);
+
+    if (!sleCred)
+        return tecNO_ENTRY;
+
+    return tesSUCCESS;
+}
+
+TER
+CredentialDelete::deleteSLE(
+    ApplyView& view,
+    std::shared_ptr<SLE> const& sle,
+    AccountID const& owner,
+    beast::Journal j)
+{
+    auto const sleOwner = view.peek(keylet::account(owner));
+    if (!sleOwner)
+        return tecINTERNAL;
+
+    if (!sle)
+        return tecNO_ENTRY;
+
+    // Remove object from owner directory
+    std::uint64_t const page = sle->getFieldU64(sfOwnerNode);
+    if (!view.dirRemove(keylet::ownerDir(owner), page, sle->key(), false))
+    {
+        JLOG(j.fatal()) << "Unable to delete Credential from owner.";
+        return tefBAD_LEDGER;
+    }
+
+    adjustOwnerCount(view, sleOwner, -1, j);
+    view.update(sleOwner);
+
+    // Remove object from ledger
+    view.erase(sle);
+
+    return tesSUCCESS;
+}
+
+TER
+CredentialDelete::doApply()
+{
+    auto const subj = ctx_.tx[~sfSubject];
+    auto const iss = ctx_.tx[~sfIssuer];
+
+    auto const& txCredType(ctx_.tx[sfCredentialType]);
+    Blob const credType{txCredType.begin(), txCredType.end()};
+
+    Keylet const kCred = keylet::credential(
+        subj ? *subj : account_, iss ? *iss : account_, credType);
+    std::shared_ptr<SLE> const sleCred = view().peek(kCred);
+
+    if (subj && iss && (subj != account_) && (iss != account_))
+    {
+        if (credentialCheckNotExpired(
+                sleCred, ctx_.view().info().parentCloseTime))
+            return tecNOT_EXPIRED;
```

Comment:
I think this should use the existing return `tecNO_PERMISSION` instead of creating a new tec code

Reply:
How about simplifying:

    auto const subject = ctx_.tx[~sfSubject].value_or(account_);
    auto const issuer = ctx_.tx[~sfIssuer].value_or(account_);

    Keylet const kCred =
        keylet::credential(subject, issuer, ctx_.tx.getFieldVL(sfCredentialType));
    std::shared_ptr<SLE> const sleCred = view().peek(kCred);

    if (subject != account_ && issuer != account_)
    {
        if (credentialCheckNotExpired(
                sleCred, ctx_.view().info().parentCloseTime))
            return tecNO_PERMISSION;
    }

Reply:
fixed